lib/storage/index.ts

```diff
@@ -1,3 +1,185 @@
-import WebStorage from './WebStorage';
+import * as Logger from '../Logger';
 
-export default WebStorage;
+import PlatformStorage from './platforms';
+import InstanceSync from './InstanceSync';
+import NoopProvider from './providers/NoopProvider';
+import type StorageProvider from './providers/types';
+
+let provider = PlatformStorage;
+let shouldKeepInstancesSync = false;
+let finishInitalization: (value?: unknown) => void;
+const initPromise = new Promise((resolve) => {
+    finishInitalization = resolve;
+});
+
+type Storage = {
+    getStorageProvider: () => StorageProvider;
+} & Omit<StorageProvider, 'name'>;
+
+/**
+ * Degrade performance by removing the storage provider and only using cache
+ */
+function degradePerformance(error: Error) {
+    Logger.logAlert(`Error while using ${provider.name}. Falling back to only using cache and dropping storage.`);
+    console.error(error);
+    provider = NoopProvider;
+}
+
+/**
+ * Runs a piece of code and degrades performance if certain errors are thrown
+ */
+function tryOrDegradePerformance<T>(fn: () => Promise<T> | T, waitForInitialization = true): Promise<T> {
+    return new Promise<T>((resolve, reject) => {
+        const promise = waitForInitialization ? initPromise : Promise.resolve();
+
+        promise.then(() => {
+            try {
+                resolve(fn());
+            } catch (error) {
+                // Test for known critical errors that the storage provider throws, e.g. when storage is full
+                if (error instanceof Error) {
+                    // IndexedDB error when storage is full (https://github.com/Expensify/App/issues/29403)
+                    if (error.message.includes('Internal error opening backing store for indexedDB.open')) {
+                        degradePerformance(error);
+                    }
```

Comment:
Should we also catch the error we are explicitly throwing here?

https://github.com/Expensify/react-native-onyx/blob/3df4fef6d78fbb8a986304e071403f7a9d8418c4/lib/storage/providers/IDBKeyValProvider.ts#L22

Reply:
@marcaaron good catch! Yes, since we can not open a connection to DB, so all calls to set/get/remove will throw exceptions.

So for me it makes sense to substitute a storage in order not to crash the app 👍